File: C++/300.cpp
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
    public:
        int lengthOfLIS(vector<int>& nums) {
            vector<int> dp(nums.size());
            int len = 0;
            for (int num : nums)
            {
                int l = 0, r = len;
                while (l != r)
                {
                    int mid = (r + l) / 2;
                    if (dp[mid] < num)
                        l = mid + 1;
                    else
                        r = mid;
                }
                dp[l] = num;
                if (l == len)
                    len++;
            }
            return len;
        }
};
__________________________________________________________________________________________________
sample 8384 kb submission
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        if(nums.empty()) return 0;
        vector<int> dp(nums.size(),1);
        int res=0;
        for(int i=0;i<nums.size();i++){
            for(int j=0;j<i;j++){
                if(nums[i]>nums[j]){
                    dp[i]=max(dp[i],dp[j]+1);
                }
            }
            res=max(res,dp[i]);
        }
        return res;
    }
};
__________________________________________________________________________________________________

File: Java/300.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public int lengthOfLIS(int[] nums) {
        int[] dp = new int[nums.length + 1];
        int size = 0;
        dp[0] = Integer.MIN_VALUE;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > dp[size]) {
                dp[++size] = nums[i];
            } else {
                int idx = binarySearch(dp, size, nums[i]);
                dp[idx] = nums[i];
            }
        }
        return size;
    }
    
    private int binarySearch(int[] a, int size, int target) {
        int lo = 0;
        int hi = size;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (a[mid] >= target) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return lo;
    }
}
__________________________________________________________________________________________________
sample 33980 kb submission
class Solution {
    public int lengthOfLIS(int[] nums) {
        int[] minLast = new int[nums.length + 1];
        Arrays.fill(minLast, Integer.MAX_VALUE);
        minLast[0] = Integer.MIN_VALUE;
        for(int i = 0; i < nums.length; i++) {
            int index = binarySearch(minLast, nums[i]);
            minLast[index] = nums[i];
        }
        for(int i = minLast.length - 1; i >= 1; i--) {
            if(minLast[i] != Integer.MAX_VALUE)
                return i;
        }
        return 0;
    }
    private int binarySearch(int[] minLast, int target) {
        int left = 0, right = minLast.length - 1;
        while(left + 1 < right) {
            int mid = (right - left) / 2 + left;
            if(minLast[mid] < target) {
                left = mid;
            } else {
                right = mid;
            }
        }
        return right;
    }
}
__________________________________________________________________________________________________
